### ISSUE TITLE
Two tasks blocked on the same file descriptor -> crash

### ISSUE THREAD

**Issue by njsmith** [dabeaz/curio#104]
While studying `_trap_io` in kernel.py, I came to two conclusions:
* I'm not clever enough to understand the `task._last_io` logic used to debounce selector registrations
* The way kernel.py uses its selector is fundamentally broken :-(. selector registrations are keyed off of the file descriptor only, so if you have two tasks that are interested in the same file descriptor then their registrations collide and things blow up.

For example, on current curio this test crashes:

```python
def test_read_write_on_same_socket(kernel):
    async def main():
        s1, s2 = socketpair()
        t1 = await spawn(s1.recv(1000))
        # Large enough send to trigger blocking on write:
        N = 10000000
        t2 = await spawn(s1.sendall(b"x" * N))
        # Above is the actual test -- right now it triggers a crash.
        # Rest of this is just to clean up:
        # Let t1 finish:
        await s2.sendall(b"x")
        # Let t2 finish:
        n = 0
        while n < N:
            n += len(await s2.recv(N))
        await t1.join()
        await t2.join()

    kernel.run(main())
```

So I thought ok, I'll fix both of these at the same time by making a new class that will sit between the kernel and the selector, and take care of multiplexing multiple waiters onto the same file descriptor, and also do debouncing.

You can see the code here: https://github.com/njsmith/curio/commit/7ece22e29089acc7efb4427608a9927de008e79d

(NB that particular commit is dangling off of some other random cleanups, so you wouldn't want to merge it directly, but it can be cherrypicked or just looked at.)

Two problems though:

1) It's slower -- the echo server microbenchmark on my laptop drops from ~40k/s to ~24k/s. There's part of me that thinks this is a good thing if it encourages us to stop paying attention to this microbenchmark, but it is nonetheless a thing...

2) It doesn't actually work in general, because the debouncing logic is too aggressive. What happens is:

   1. Some task is interested in fd 4
   2. Then it stops being interested in fd 4
   3. We make a note that if no-one has become interested in fd 4 by the
      time the next loop rolls around, then we should drop it from the
      underlying selector.
   4. Then our task closes fd 4.
   5. Then it (or some other task) opens a new file, which gets assigned fd 4.
   6. Then it blocks on this new file.
   7. The debouncing logic sees that it's still interested in fd 4, and it "knows" the selector is still watching fd 4, so it doesn't adjust the selector.
   8. But, epoll is keyed off actual files, not file descriptor
      numbers. It knows that these two fd 4s are not the same. In fact,
      it already removed fd 4 from its watch set when the first file
      descriptor was closed. So we are screwed.

But, looking at this... actually the same basic problem could happen to the current debouncing logic. The reason it doesn't is that the new debouncing logic is keyed off of file descriptors, while the old debouncing logic is keyed off of file objects. But you kind of *have* to key off of file descriptors if you want to handle the general case of arbitrarily many tasks that are interested in the same file descriptor, because the underlying selector logic is keyed off of file descriptors -- if you try to register two separate interests in the same fd but using different fileobjs, then the selector will blow up.

So, maybe we should just give up on debouncing?

It looks like asyncio avoids some of these issues by imposing the rule that there can only be one reader and one writer per file descriptor at any given time: https://github.com/python/asyncio/blob/eb4c3ffdddc415330b434cefa9187360dc934ce0/asyncio/selector_events.py#L260-L273

In practice, I can't imagine any situations where it would make sense to have two tasks both blocked on the readability or both blocked on the writeability of the same fd, so I guess this wouldn't be a terribly onerous restriction. Not sure how much imposing it simplifies things, though.

**Comment by njsmith** [comment]
BTW, do you remember what evidence you had for adding the debouncing in the first place?

**Comment by dabeaz** [comment]
The rationale for the debouncing is that if a task reads once, there's a pretty good chance it's going to read again.   For example, reading some headers on a request and then reading a request body.  If a task writes, there's a good chance it's going to write again.  For example, writing a large response in chunks.  I consider both of these cases to be real-world scenarios--not just something from an echo server. 

I consider any situation where two tasks are blocked on the same descriptor to be an error.  It's already an error in the selectors module so far as I can tell.  Real screwy things happen if you do things like this with threads as well (at least in my experience).     If someone really wants multiple tasks to be doing something with the same descriptor, they should probably use os.dup() on it and work with that.  

The case of closing/reopening should be looked at though.  That is a bug.  

As a side-note: some other async frameworks seem to make a similar "optimization" by simply leaving the file descriptor attached to the event loop and performing their own internal buffering.  This ties it to the buffer-bloat problem.  There is peril everywhere. 

I will ponder this for a bit...

**Comment by njsmith** [comment]
I understand the idea of the debouncing; my question is more, like, do we
expect this to be a 1% or 10% or 100% speedup? Or in principle it could
even be a slowdown if it takes longer to compute the debounce information
than it does to just unregister+reregister.
.
And agreed that having multiple readers or multiple writers is weird, but
one task reading from a file descriptor while another task writes to that
same fd is *really* normal. See e.g. that bidirectional proxy example I
used to motivate defer_cancellation. I can't imagine requiring a dup for
this case.

**Comment by dabeaz** [comment]
The debounce computation is extremely trivial compared to the expense of deregistering from the event loop, making system calls to the kernel, and all of the overhead associated with that.  I'd be sad to lose the extra bit of performance gain, but I'm not married to the debouncing code either.

At the moment, curio is constrained to having a single read or write pending on each fd.  This is mainly a limitation of the selectors module.  However, I just spent a few moments poking around in asyncio internals, it seems that we could probably support both a pending read and a write without much trouble.    

Let me poke around on this.

**Comment by dabeaz** [comment]
I have pushed a different kind of change that allows two different tasks to read/write on the same file descriptor.  The test code above passes with this change.   The debouncing logic is still being used (but is refined).    I still need to look at the possible case of duplicate file descriptors (reuse) messing things up. 

All of the tests pass, but this would definitely need more hammering on.    The change does NOT allow more than one task to read on the same file or to write on the same file at the same time (i.e., two tasks both expressing an interest in reading).   I'm not exactly sure how common it would be to write code like that. 

That said, do any of the other async frameworks actually allow multiple readers/writers on the same file descriptor?   I looked inside asyncio and it doesn't seem like it would allow such a configuration.    However, I know from operating systems that you can most definitely have multiple threads/processes trying to read/write on the same file descriptor at once.    Maybe there's something interesting for curio to explore there.    I'm not sure.  Just thinking out loud.

**Comment by dabeaz** [comment]
Just wondering... having reviewed the code with the new selector class, it seems like it wakes all waiting tasks whenever there is a matching I/O event.   This could create a real interesting sort of thundering herd problem.  For example, maybe the first task only does a partial read of the data (because of the maxsize on recv).  But then the 2nd task picks up where it left off and reads any remaining data on the socket.  The 2nd task would do this without any sort of blocking since the data would just be sitting there, ready to be read.    Meh.  I don't know.  Maybe it's better to limit this to one task reading, one task writing, per file descriptor.

**Comment by dabeaz** [comment]
Just a note:  I'm hoping to look at the reused file descriptor part of this bug report.   Been completely distracted over the last few weeks with training courses. May have more time over the upcoming holiday.

**Comment by dabeaz** [comment]
I've pushed a change that addresses the problem of file descriptor reuse.  It turns out that you can (apparently) easily solve this problem by putting a small wrapper object around the integer file descriptor and passing that into Curio instead.  Although the integer file descriptor value might be reused, the wrapper object is not.  So, it becomes possible to detect changes to the file via that route.   I also made the kernel NOT apply the debouncing optimization if bare integer file descriptors are used--an extra layer of safety perhaps.

If at all possible, I'd like to try and keep the so-called debouncing optimization.  Benchmarking some stuff while working on this, it makes a 25% performance difference.  That's not entirely insignificant.   

Writing tests for this is surprising hard.   Still need to add some cleaned up unit tests that check for duplicate file descriptor reuse as described in this issue.

**Comment by ZhukovAlexander** [comment]
@njsmith , should we close this due to 	792988c?

**Comment by njsmith** [comment]
Depends on whether @dabeaz thinks he's done or not :-)

**Comment by dabeaz** [comment]
I don't mind closing the issue, but wonder if there is more discussion.

**Comment by dabeaz** [comment]
Now I know why I left this open... still thinking about the case of two tasks reading on the same file descriptor.   Behavior is currently undefined.  Maybe it should raise an exception.

**Comment by dabeaz** [comment]
Modified to raise an exception if two tasks attempt to wait on the same file descriptor at once.
